core/src/transport/mod.rs

```diff
@@ -54,6 +54,36 @@ pub use self::memory::connector;
 pub use self::muxed::MuxedTransport;
 pub use self::upgrade::UpgradedNode;
 
+
+// This is a new error type that you've created. It represents the ways a
+// toolchain could be invalid.
+//
+// The custom derive for Fail derives an impl of both Fail and Display.
+// We don't do any other magic like creating new types.
+#[derive(Debug, Fail)]
+pub enum TransportError {
+    #[fail(display= "Error parsing Address {}: {}", addr, err)]
+    MultiAddrParseError {
+        addr: Multiaddr,
+        #[cause] err: MultiaddrError
+    },
+    #[fail(display = "Listening is not support for Multiaddress {}", _0)]
+    ListenNotSupported(Multiaddr),
+
+    #[fail(display = "Listening on Multiaddress {} failed: {}", _0, _1)]
+    ListenFailed(Multiaddr, #[cause] IoError),
+
+    #[fail(display = "Dialing is not support for Multiaddress {}", _0)]
+    DialNotSupported(Multiaddr),
+
+    #[fail(display = "Dialing for Multiaddress {} failed: {}", _0, _1)]
+    DialingFailed(Multiaddr, #[cause] IoError),
+}
+
+pub type ListenerResult<T> = Result<(<T as Transport>::Listener, Multiaddr), (T, TransportError)>;
```

Comment:
I'm not a fan of these typedefs, because when I see `FooResult<T>` I expect that the `Ok` will contain a `T` and not something else.

Reply:
Also, for the error I think that the `T` should be directly inside of `TransportError`.

The `T` wouldn't be printed on the `Debug` and `Display` implementations of `TransportError`.
Most often than not, we don't want to retrieve the `T`. Having it inside the `TransportError` means that we can just do `dial(addr).unwrap()` without having to implement `Debug` on `T`.

Reply:
> I'm not a fan of these typedefs, because when I see `FooResult<T>` I expect that the `Ok` will contain a `T` and not something else.

I could switch it to `ListenerResult<Self::Listener, Self>`, which hits closer to home, but that is redundant. I intended on removing the `Self` in there altogether, but that isn't possible on the `trait`, you'd still have to spell it out on every implementation.

> Also, for the error I think that the `T` should be directly inside of `TransportError`.
Agreed. And intended to do just that. However, we have the issue that `failure` requires all its fields to be `Send` + `Sync` to provide stable multi-threading behavior of the error (which makes sense to me, tbh), but our `Transport`s aren't.

---

IMHO, both of these highlight a different aspect though: that having `Self` be a result of the error is an anti-pattern, that we shouldn't even be consuming `self` in these functions in the first place is.

Reply:
> I could switch it to ListenerResult<Self::Listener, Self>

It should be just be `Result<Self::Dial, TransportError<T>>` and `Result<(Self::Listener, Multiaddr), TransportError<T>>` manually everywhere IMO. No need to have a typedef at all.

> failure requires all its fields to be Send + Sync to provide stable multi-threading behavior of the error (which makes sense to me, tbh)

I don't see any reason why errors need to be `Sync`, and even `Send` is very debatable.
It's true however that we should have an enum containing the "raw" error, and a struct containing this enum in addition to `T`.
That may not be possible with only `failure` though.

Reply:
> No need to have a typedef at all

Typedef aliases are mostly for readiblity in the first place. I don't know of any case where they are "needed" ;) .

However, through replacing the consuming self with a referencing self-system, I was able to make this into a neat little `TransportResult` shortcut, which I believe you'd find to be having the expressiveness you'd acquired:
```rust
    fn listen_on(&self, addr: Multiaddr) -> TransportResult<(Self::Listener, Multiaddr)>

    fn dial(&self, addr: Multiaddr) -> TransportResult<Self::Dial>
```